src/index.js

```diff
@@ -19,7 +19,27 @@ plugins.flow = flowPlugin;
 plugins.jsx = jsxPlugin;
 
 export function parse(input, options) {
-  return new Parser(options, input).parse();
+  if (options && options.sourceType === "unambiguous") {
+    options = Object.assign({}, options);
+    try {
+      options.sourceType = "module";
+      const ast = new Parser(options, input).parse();
+
+      // Rather than try to parse as a script first, we opt to parse as a module and convert back
+      // to a script where possible to avoid having to do a full re-parse of the input content.
+      if (!hasModuleSyntax(ast)) ast.program.sourceType = "script";
```

Comment:
For perf I had the acorn folks dig into proofing out [an optimized top level parse](https://github.com/RReverser/esmod/blob/master/index.js). In reify we also do a source sniff for `import` or `export` and if it fails we skip it.